### PR TITLE
refactor(GUI): deangular os-dialog and error module

### PR DESCRIPTION
**Pull request by Shou**
We remove the Angular dependency from `lib/gui/os/dialog` and `lib/gui/modules/error` which is also now renamed to `lib/gui/modules/exception`.

Changelog-Entry: Deangular the os-dialog and error modules.

### CODE REVIEW

File: lib/gui/app.js [hunk 1]
```diff
@@ -196,13 +194,13 @@ app.run(($timeout, ErrorService) => {
...
     driveScanner.stop();
 
-    return ErrorService.reportException(error);
+    return exception.report(error);
   });
 
   driveScanner.start();
 });
 
-app.run(($window, WarningModalService, ErrorService, OSDialogService) => {
```
Comment (jviotti) [inline lib/gui/app.js]: Good catch!

File: lib/gui/app.js [hunk 2]
```diff
@@ -264,11 +262,10 @@ app.config(($urlRouterProvider) => {
 });
 
 app.config(($provide) => {
-  $provide.decorator('$exceptionHandler', ($delegate, $injector) => {
-    return (exception, cause) => {
-      const ErrorService = $injector.get('ErrorService');
-      ErrorService.reportException(exception);
-      $delegate(exception, cause);
+  $provide.decorator('$exceptionHandler', ($delegate) => {
+    return (except, cause) => {
```
Comment (jviotti) [inline lib/gui/app.js]: Hm, I wonder if this means we should pick another name for the `exception` module. What about `exceptionReporter`?

File: lib/gui/os/dialog.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+ * @fulfil {Object} - selected image
+ * @returns {Promise};
+ *
+ * @example
+ * osDialog.selectImage().then((image) => {
+ *   console.log('The selected image is', image.path);
+ * });
+ */
+exports.selectImage = () => {
+  return new Promise((resolve) => {
```
Comment (jviotti) [inline lib/gui/os/dialog.js]: Can we use `Bluebird` here? I remember there were some very subtle differences between ES6 Promises and Bluebird's that caused use some headaches before.

File: lib/gui/os/dialog.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+ * @public
+ *
+ * @description
+ * Notice that by image, we mean *.img/*.iso/*.zip/etc files.
+ *
+ * @fulfil {Object} - selected image
+ * @returns {Promise};
+ *
+ * @example
+ * osDialog.selectImage().then((image) => {
```
Comment (jviotti) [inline lib/gui/os/dialog.js]: Should we call this module simply `dialog` now?

File: lib/gui/pages/main/controllers/image-selection.js [hunk 5]
```diff
@@ -138,17 +139,19 @@ module.exports = function(
   this.openImageSelector = () => {
     analytics.logEvent('Open image selector');
 
-    OSDialogService.selectImage().then((imagePath) => {
+    osDialog.selectImage().then((imagePath) => {
+      $timeout(() => {
```
Comment (jviotti) [inline lib/gui/pages/main/controllers/image-selection.js]: Is this timeout needed to trigger the digest loop? If so, should we only add it around `this.selectImageByPath(imagePath)`?

File: lib/gui/os/dialog.js (hunk 4 above)
Comment (Shou) [inline lib/gui/os/dialog.js]: It makes it more susceptible to a name clash like `exception` so I think let's keep it for now.

**Comment by Shou** [comment]
What do you think of this now?

File: lib/gui/app.js [hunk 6]
```diff
@@ -42,6 +42,8 @@ const updateNotifier = require('./components/update-notifier');
 const availableDrives = require('./models/available-drives');
 const selectionState = require('./models/selection-state');
 const driveScanner = require('./modules/drive-scanner');
+const osDialog = require('./os/dialog');
+const exceptionReporter = require('./modules/exception');
```
Comment (jviotti) [inline lib/gui/app.js]: Can we also rename the file itself to `exception-reporter`?

File: lib/gui/pages/main/controllers/image-selection.js [hunk 7]
```diff
@@ -122,20 +123,22 @@ module.exports = function(
...
-    imageStream.getImageMetadata(imagePath)
-      .then(this.selectImage)
-      .catch((error) => {
-        const imageError = errors.createUserError({
-          title: 'Error opening image',
-          description: messages.error.openImage({
-            imageBasename: path.basename(imagePath),
-            errorMessage: error.message
-          })
+    $timeout(() => {
```
Comment (jviotti) [inline lib/gui/pages/main/controllers/image-selection.js]: As far as I can see, the `$timeout` is only needed for `this.selectImage`. If so, can we do

```js
imageStream.getImageMetadata(imagePath)
    .then((imageMetadata) => {
        $timeout(() => {
            this.selectImage(imageMetadata);
        });
    });
```

?

File: lib/gui/app.js (hunk 6 above)
Comment (Shou) [inline lib/gui/app.js]: Sure, done.

File: lib/gui/pages/main/controllers/image-selection.js (hunk 7 above)
Comment (Shou) [inline lib/gui/pages/main/controllers/image-selection.js]: Lol my bad, I don't know what I was doing wrapping a promise in the timeout. I've fixed it now.

File: lib/gui/app.js (hunk 2 above)
Comment (jviotti) [inline lib/gui/app.js]: Can we now rename this variable to `exception`?

**Comment by Shou** [comment]
Done!

File: lib/gui/app.js [hunk 8]
```diff
@@ -264,11 +262,10 @@ app.config(($urlRouterProvider) => {
 });
 
 app.config(($provide) => {
-  $provide.decorator('$exceptionHandler', ($delegate, $injector) => {
+  $provide.decorator('$exceptionHandler', ($delegate) => {
     return (exception, cause) => {
-      const ErrorService = $injector.get('ErrorService');
-      ErrorService.reportException(exception);
-      $delegate(exception, cause);
+      exceptionReporter.report(except);
```
Comment (jviotti) [inline lib/gui/app.js]: I think you forgot to update the old references to `except` :)

**Comment by jviotti** [comment]
I'm rebasing this myself, to see if I can get it merged soon.

**Comment by lurch** [comment]
nit: commit-message refers to `lib/gui/modules/exception` but it actually ended up getting named `lib/gui/modules/exception-reporter.js`

**Pull request merged**